File: Problems/0382. (M) Linked List Random Node/Java/solution.java
class Solution {
  // naive solution, traverse list twice
  int length = 0;
  ListNode head;
  // Java needs to instantiate object to use prng methods
  Random prng;

  public Solution(ListNode head) {
    this.head = head;
    this.prng = new Random();
    ListNode cur = head;
    // determine length of list
    while (cur.next != null) {
      this.length++;
      cur = cur.next;
    }
  }
  
  public int getRandom() {
    // retrieve random nth node in list
    // nextInt(n) generates int in range [0, n)
    // this.length is 0-indexed so add 1 to n
    int n = this.prng.nextInt(this.length+1);
    ListNode cur = this.head;
    while (n > 0 && cur.next != null) {
      cur = cur.next;
      n--;
    }
    return cur.val;
  }
}

File: Problems/0382. (M) Linked List Random Node/Python3/solution.py
class Solution:
  # naive solution, traverse list twice

  def __init__(self, head: Optional[ListNode]):
    self.length = 1
    self.head = head
    cur = head
    # determine length of list
    while cur.next:
      cur = cur.next
      self.length += 1

  def getRandom(self) -> int:
    # retrieve random nth node in list
    n = random.randrange(1, self.length+1)
    cur = self.head
    while n > 1 and cur.next:
      cur = cur.next
      n -= 1
    
    return cur.val

